libp2p/transports/tls/certificate.nim

```diff
@@ -0,0 +1,489 @@
+  len += retSig
+
+  # Write publicKey OCTET STRING
+  let retPub = mbedtls_asn1_write_octet_string(
+    addr extPtr, startPtr, unsafeAddr pubKey[0], pubKey.len.uint
+  )
+  if retPub < 0:
+    raise newException(ASN1EncodingError, "Failed to write publicKey OCTET STRING")
+  len += retPub
+
+  # Total length of the SEQUENCE contents
+  let contentLen = retSig + retPub
+  # Write SEQUENCE length
+  let retLen = mbedtls_asn1_write_len(addr extPtr, startPtr, contentLen.uint)
+  if retLen < 0:
+    raise newException(ASN1EncodingError, "Failed to write SEQUENCE length")
+  len += retLen
+
+  # Write SEQUENCE tag
+  let retTag = mbedtls_asn1_write_tag(
+    addr extPtr, startPtr, MBEDTLS_ASN1_CONSTRUCTED or MBEDTLS_ASN1_SEQUENCE
+  )
+  if retTag < 0:
+    raise newException(ASN1EncodingError, "Failed to write SEQUENCE tag")
+  len += retTag
+
+  # Calculate dataOffset based on the accumulated length
+  let dataOffset = extValueSize - len - 1
+
+  # Extract the relevant portion of extValue as a seq[byte]
+  let extValueSeq = toSeq(extValue[dataOffset ..< extValueSize])
+
+  # Return the extension content
+  return extValueSeq
+
+proc makeLibp2pExtension(
+    identityKeypair: KeyPair, certificateKeypair: mbedtls_pk_context
+): seq[byte] {.
+    raises: [
+      CertificateCreationError, IdentityPubKeySerializationError, IdentitySigningError,
+      ASN1EncodingError, TLSCertificateError,
+    ]
+.} =
+  ## Creates the libp2p extension containing the SignedKey.
+  ##
+  ## The libp2p extension is an ASN.1-encoded structure that includes
+  ## the public key and its signature over the certificate's public key.
+  ##
+  ## Parameters:
+  ## - `identityKeypair`: The peer's identity key pair.
+  ## - `certificateKeypair`: The key pair used for the certificate.
+  ##
+  ## Returns:
+  ## A sequence of bytes representing the libp2p extension.
+  ##
+  ## Raises:
+  ## - `CertificateCreationError` if public key serialization fails.
+  ## - `IdentityPubKeySerializationError` if serialization of identity public key fails.
+  ## - `IdentitySigningError` if signing the hash fails.
+  ## - `ASN1EncodingError` if ASN.1 encoding fails.
+
+  # Serialize the Certificate's Public Key
+  var
+    certPubKeyDer: array[512, byte]
+    certPubKeyDerLen: cint
+
+  certPubKeyDerLen = mbedtls_pk_write_pubkey_der(
+    unsafeAddr certificateKeypair, addr certPubKeyDer[0], certPubKeyDer.len.uint
+  )
+  if certPubKeyDerLen < 0:
+    raise newException(
+      CertificateCreationError, "Failed to write certificate public key in DER format"
+    )
+
+  # Adjust pointer to the start of the data
+  let certPubKeyDerPtr = addr certPubKeyDer[certPubKeyDer.len - certPubKeyDerLen]
+
+  # Create the Message to Sign
+  var msg = newSeq[byte](P2P_SIGNING_PREFIX.len + certPubKeyDerLen.int.int)
+
+  # Copy the prefix into msg
+  for i in 0 ..< P2P_SIGNING_PREFIX.len:
+    msg[i] = byte(P2P_SIGNING_PREFIX[i])
```

Comment:
Maybe use `copyMem`?

Reply:
Seems it's not possible to use it with a `const`